File: go/kdb/integrity/repair_test.go
package integrity

import (
	"testing"

	"github.com/limidus/kdb/go/kdb/storage"
	storio "github.com/limidus/kdb/go/kdb/storage/io"
)

func TestRepairTornTailTruncatesAndQuarantines(t *testing.T) {
	shim := newTestShim(t)
	const ns = "ns1"
	commits := buildChain(t, 2, ns)
	full := rawFrame(t, commits[0])
	torn := rawFrame(t, commits[1])[:10]
	appendSegment(t, shim, ns, 0, full, torn)

	opts := Options{Level: L1, Compression: storage.CompressionNone}
	report, err := Verify(shim, ns, opts)
	if err != nil {
		t.Fatal(err)
	}
	result, err := Repair(shim, report, opts)
	if err != nil {
		t.Fatal(err)
	}
	if len(result.Steps) != 1 || result.Steps[0].Action != ActionTruncatedTornTail {
		t.Fatalf("expected one truncate step, got %+v", result.Steps)
	}
	if result.Steps[0].QuarantineName == "" {
		t.Fatal("expected a quarantine file name")
	}

	quarantined, err := shim.ReadFromSegment(result.Steps[0].QuarantineName, 0, 1<<20)
	if err != nil {
		t.Fatalf("quarantine file not readable: %v", err)
	}
	if string(quarantined) != string(torn) {
		t.Fatal("quarantined bytes must be exactly the removed torn tail")
	}

	report2, err := Verify(shim, ns, opts)
	if err != nil {
		t.Fatal(err)
	}
	if !report2.Clean() {
		t.Fatalf("expected clean report after repair, got %+v", report2.Findings)
	}
	if report2.Segments[0].FrameCount != 1 {
		t.Fatalf("expected 1 frame surviving repair, got %+v", report2.Segments[0])
	}
}

func TestRepairIsIdempotent(t *testing.T) {
	shim := newTestShim(t)
	const ns = "ns1"
	commits := buildChain(t, 2, ns)
	appendSegment(t, shim, ns, 0, rawFrame(t, commits[0]), rawFrame(t, commits[1])[:10])

	opts := Options{Level: L1, Compression: storage.CompressionNone}
	report, err := Verify(shim, ns, opts)
	if err != nil {
		t.Fatal(err)
	}
	if _, err := Repair(shim, report, opts); err != nil {
		t.Fatal(err)
	}

	report2, err := Verify(shim, ns, opts)
	if err != nil {
		t.Fatal(err)
	}
	result2, err := Repair(shim, report2, opts)
	if err != nil {
		t.Fatal(err)
	}
	if len(result2.Steps) != 0 {
		t.Fatalf("expected no-op on an already-repaired report, got %+v", result2.Steps)
	}
}

func TestRepairMidLogRemovesOnlyProvenSafeFrame(t *testing.T) {
	shim := newTestShim(t)
	const ns = "ns1"
	c0 := buildCommit(t, ns, nil)
	c1 := buildCommit(t, ns, &c0.Hash)  // kept: precedes the corrupt frame in its segment
	c1b := buildCommit(t, ns, &c0.Hash) // corrupted: a dead end nothing else references
	c2 := buildCommit(t, ns, &c0.Hash)  // independent, in a later segment

	appendSegment(t, shim, ns, 0, rawFrame(t, c0))
	appendSegment(t, shim, ns, 1, rawFrame(t, c1), flippedFrame(t, c1b))
	appendSegment(t, shim, ns, 2, rawFrame(t, c2))

	opts := Options{Level: L1, Compression: storage.CompressionNone}
	report, err := Verify(shim, ns, opts)
	if err != nil {
		t.Fatal(err)
	}
	if !hasFinding(report, MidLogCorruption, 1) {
		t.Fatalf("expected mid_log_corruption at segment 1, got %+v", report.Findings)
	}

	result, err := Repair(shim, report, opts)
	if err != nil {
		t.Fatal(err)
	}
	if len(result.Steps) != 1 || result.Steps[0].Action != ActionRewroteSegmentPrefix {
		t.Fatalf("expected one prefix-rewrite step, got %+v", result.Steps)
	}

	report2, err := Verify(shim, ns, Options{Level: L2, Compression: storage.CompressionNone})
	if err != nil {
		t.Fatal(err)
	}
	if !report2.Clean() {
		t.Fatalf("expected clean report after safe repair, got %+v", report2.Findings)
	}
}

func TestRepairRefusesWhenClosureBreaks(t *testing.T) {
	shim := newTestShim(t)
	const ns = "ns1"
	c0 := buildCommit(t, ns, nil)
	c1 := buildCommit(t, ns, &c0.Hash) // corrupted, alone in its segment
	c2 := buildCommit(t, ns, &c1.Hash) // depends on the corrupted commit

	appendSegment(t, shim, ns, 0, rawFrame(t, c0))
	appendSegment(t, shim, ns, 1, flippedFrame(t, c1))
	appendSegment(t, shim, ns, 2, rawFrame(t, c2))

	opts := Options{Level: L1, Compression: storage.CompressionNone}
	report, err := Verify(shim, ns, opts)
	if err != nil {
		t.Fatal(err)
	}

	name := storio.SegmentNameBuilder.DeltaSequenced(ns, 1)
	before, err := shim.ReadFromSegment(name, 0, 1<<20)
	if err != nil {
		t.Fatal(err)
	}

	result, err := Repair(shim, report, opts)
	if err != nil {
		t.Fatal(err)
	}
	if len(result.Steps) != 1 || result.Steps[0].Action != ActionRefused {
		t.Fatalf("expected a refused step, got %+v", result.Steps)
	}
	if len(result.Steps[0].MissingHashes) != 1 || result.Steps[0].MissingHashes[0] != c1.Hash.Hex() {
		t.Fatalf("expected missing hash %s, got %+v", c1.Hash.Hex(), result.Steps[0].MissingHashes)
	}

	after, err := shim.ReadFromSegment(name, 0, 1<<20)
	if err != nil {
		t.Fatal(err)
	}
	if string(before) != string(after) {
		t.Fatal("a refused repair must not touch the segment")
	}
}

File: kdb-integrity/src/commonTest/kotlin/dev/kdb/integrity/RepairTest.kt
package dev.kdb.integrity

import dev.kdb.storage.CompressionCodec
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue
import kotlinx.coroutines.test.runTest

class RepairTest {
    @Test
    fun tornTailTruncatesAndQuarantines() =
        runTest {
            val shim = newTestShim()
            val ns = "ns1"
            val commits = buildChain(2, ns)
            val full = rawFrame(commits[0])
            val torn = rawFrame(commits[1]).copyOfRange(0, 10)
            appendSegment(shim, ns, 0, full, torn)

            val opts = Options(Level.L1, CompressionCodec.NONE)
            val report = verify(shim, ns, opts)
            val result = repair(shim, report, opts)
            assertEquals(1, result.steps.size)
            assertEquals(Action.TRUNCATED_TORN_TAIL, result.steps[0].action)
            assertTrue(result.steps[0].quarantineName.isNotEmpty())

            val quarantined = shim.readFromSegment(result.steps[0].quarantineName, 0, 1 shl 20)
            assertTrue(quarantined.contentEquals(torn), "quarantined bytes must be exactly the removed torn tail")

            val report2 = verify(shim, ns, opts)
            assertTrue(report2.isClean, "expected clean report after repair, got ${report2.findings}")
            assertEquals(1, report2.segments[0].frameCount)
        }

    @Test
    fun repairIsIdempotent() =
        runTest {
            val shim = newTestShim()
            val ns = "ns1"
            val commits = buildChain(2, ns)
            appendSegment(shim, ns, 0, rawFrame(commits[0]), rawFrame(commits[1]).copyOfRange(0, 10))

            val opts = Options(Level.L1, CompressionCodec.NONE)
            repair(shim, verify(shim, ns, opts), opts)

            val result2 = repair(shim, verify(shim, ns, opts), opts)
            assertEquals(0, result2.steps.size, "expected no-op on an already-repaired report, got ${result2.steps}")
        }

    @Test
    fun repairMidLogRemovesOnlyProvenSafeFrame() =
        runTest {
            val shim = newTestShim()
            val ns = "ns1"
            val c0 = buildCommit(ns, null)
            val c1 = buildCommit(ns, c0.hash) // kept: precedes the corrupt frame in its segment
            val c1b = buildCommit(ns, c0.hash) // corrupted: a dead end nothing else references
            val c2 = buildCommit(ns, c0.hash) // independent, in a later segment

            appendSegment(shim, ns, 0, rawFrame(c0))
            appendSegment(shim, ns, 1, rawFrame(c1), flippedFrame(c1b))
            appendSegment(shim, ns, 2, rawFrame(c2))

            val opts = Options(Level.L1, CompressionCodec.NONE)
            val report = verify(shim, ns, opts)
            assertTrue(hasFinding(report, Classification.MID_LOG_CORRUPTION, 1), "expected mid_log_corruption at segment 1, got ${report.findings}")

            val result = repair(shim, report, opts)
            assertEquals(1, result.steps.size)
            assertEquals(Action.REWROTE_SEGMENT_PREFIX, result.steps[0].action)

            val report2 = verify(shim, ns, Options(Level.L2, CompressionCodec.NONE))
            assertTrue(report2.isClean, "expected clean report after safe repair, got ${report2.findings}")
        }

    @Test
    fun repairRefusesWhenClosureBreaks() =
        runTest {
            val shim = newTestShim()
            val ns = "ns1"
            val c0 = buildCommit(ns, null)
            val c1 = buildCommit(ns, c0.hash) // corrupted, alone in its segment
            val c2 = buildCommit(ns, c1.hash) // depends on the corrupted commit

            appendSegment(shim, ns, 0, rawFrame(c0))
            appendSegment(shim, ns, 1, flippedFrame(c1))
            appendSegment(shim, ns, 2, rawFrame(c2))

            val opts = Options(Level.L1, CompressionCodec.NONE)
            val report = verify(shim, ns, opts)

            val name = "ns/$ns/delta/00000000000000000001.seg"
            val before = shim.readFromSegment(name, 0, 1 shl 20)

            val result = repair(shim, report, opts)
            assertEquals(1, result.steps.size)
            assertEquals(Action.REFUSED, result.steps[0].action)
            assertEquals(listOf(c1.hash.toHex()), result.steps[0].missingHashes)

            val after = shim.readFromSegment(name, 0, 1 shl 20)
            assertTrue(before.contentEquals(after), "a refused repair must not touch the segment")
        }
}
